File: 0865_Smallest Subtree with all the Deepest Nodes/SmallestSubtreewithalltheDeepestNodes.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def subtreeWithAllDeepest(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        def traverse(node):
            if not node:return (None,0)
            leftNode, leftH = traverse(node.left)
            rightNode, rightH = traverse(node.right)
            if(leftH == rightH):
                return (node, leftH+1)
            elif(leftH > rightH):
                return (leftNode, leftH+1)
            else:
                return (rightNode, rightH+1)
        
        return traverse(root)[0]

File: 0865_Smallest Subtree with all the Deepest Nodes/SmallestSubtreewithalltheDeepestNodes.java
class Solution {
    private int max = -1;
    private TreeNode ans;
    public TreeNode subtreeWithAllDeepest(TreeNode root) {
        dfs(root, 0);
        return ans;
    }
    
    private int dfs(TreeNode node, int depth) {
        if(node == null) return depth - 1;
        int left = dfs(node.left, depth+1);
        int right = dfs(node.right, depth + 1);
        if(depth > max) max = depth;
        if(left == max && right == max) ans = node;
        return Math.max(left, right);
    }
}